### ISSUE TITLE
Add more details to error if gem version in `Gemfile.lock` is not compatible with current ruby version

### ISSUE THREAD

**Issue by ShockwaveNN** [rubygems/rubygems#5435]
### Describe the problem as clearly as you can

This issue is a result of https://github.com/grosser/parallel_tests/issues/846#issuecomment-1080782798

`parallel_tests` dropped support of `ruby-2.6` in `parallel_tests-2.8.0`
But if you Gemfile.lock frozen on newer version error message for `bundler install` looks like this version is `yanked`, not like `your ruby is not supporting this version of gem`

### Did you try upgrading rubygems & bundler?

Yes, latest version 

### Post steps to reproduce the problem

1. Create `Gemfile` with content:
```
source 'https://rubygems.org'

gem "parallel_tests"
```
2. Create `Gemfile.lock` with `parallel_tests` locked on 3.8.0 (which do not support ruby-2.6):
```
GEM
  remote: https://rubygems.org/
  specs:
    parallel (1.22.1)
    parallel_tests (3.8.0)
      parallel

PLATFORMS
  x86_64-linux

DEPENDENCIES
  parallel_tests

BUNDLED WITH
   2.3.10

```
3. Create `Dockerfile`:
```
FROM ruby:2.6

RUN gem update --system
ADD Gemfile .
ADD Gemfile.lock .
RUN bundle install

```

4. run `docker build .`

### What were you expecting to happen?

Message like
```
You've trying to install `parallel_tests` on `ruby-2.6`, but `parallel_tests` has `.required_ruby_version = '>= 2.7.0'` in its sourced
```

### What actually happened?

```
Your bundle is locked to parallel_tests (3.8.0) from rubygems repository
https://rubygems.org/ or installed locally, but that version can no longer be
found in that source. That means the author of parallel_tests (3.8.0) has
removed it. You'll need to update your bundle to a version other than
parallel_tests (3.8.0) that hasn't been removed in order to install.
```

The problem with this message, that it's looked like `parallel_tests (3.8.0) ` was yanked
I've spend a hour understanding why my CI on `ruby-2.6` is failing and not failing on `ruby-3.1` and how one gem may be yanked only for some ruby version

## Environment

```
Bundler       2.3.10
  Platforms   ruby, x86_64-linux
Ruby          2.6.9p207 (2021-11-24 revision 67954) [x86_64-linux]
  Full Path   /usr/local/bin/ruby
  Config Dir  /usr/local/etc
RubyGems      3.3.10
  Gem Home    /usr/local/bundle
  Gem Path    /root/.gem/ruby/2.6.0:/usr/local/lib/ruby/gems/2.6.0:/usr/local/bundle
  User Home   /root
  User Path   /root/.gem/ruby/2.6.0
  Bin Dir     /usr/local/bundle/bin
Tools         
  Git         2.30.2
  RVM         not installed
  rbenv       not installed
  chruby      not installed
```

## Bundler Build Metadata

```
Built At          2022-03-23
Git SHA           4bbb70e7de
Released Version  true
```

## Bundler settings

```
app_config
  Set via BUNDLE_APP_CONFIG: "/usr/local/bundle"
silence_root_warning
  Set via BUNDLE_SILENCE_ROOT_WARNING: true
```

### Gemfile

```ruby
source 'https://rubygems.org'


gem "parallel_tests"
```

### Gemfile.lock

```
GEM
  remote: https://rubygems.org/
  specs:
    parallel (1.22.1)
    parallel_tests (3.8.0)
      parallel

PLATFORMS
  x86_64-linux

DEPENDENCIES
  parallel_tests

BUNDLED WITH
   2.3.10
```

**Comment by deivid-rodriguez** [comment]
Thank you, that's such an unfortunate error indeed, I will try to fix this.

**Comment by deivid-rodriguez** [comment]
Also, in non frozen mode, I would actually expect this to just work and update the `Gemfile.lock` to use an older version of `parallel_tests` that does work with Ruby 2.6.

**Comment by ShockwaveNN** [comment]
> I would actually expect this to just work and update the Gemfile.lock to use an older version of parallel_tests that does work with Ruby 2.6.

I think it will be a correct behavior but still some warning still be shown

Something like

```
[Warning] Using `parallel_tests 2.7` since it's the latest version supporting your Ruby 2.6. Consider upgrading ruby to use latest version of `parallel_tests 2.8`
```

**Comment by coldnebo** [comment]
I just ran into a similar issue with ruby 2.6.8 and bundler 2.3.11 under Debian 11:

```
Your bundle is locked to actioncable (5.1.6.1) from locally installed gems, but that version can no longer be found in that source. That means the
author of actioncable (5.1.6.1) has removed it. You'll need to update your bundle to a version other than actioncable (5.1.6.1) that hasn't been
removed in order to install.
```

I know for a fact this isn't true.  actioncable 5.1.6.1 exists and has not been yanked.

AFAIK, there is no ruby version constraint for this gem under ruby 2.6.8, can you help me where this error is coming from?  Is there a way to debug or interrogate the constraints?  

The devilish thing is that if I go through and manually gem install these same versions, everything works, but I really shouldn't have to do that.

**Comment by deivid-rodriguez** [comment]
Can you show your `Gemfile.lock` file? Note that the error message says **from locally installed gems**, not form `https://rubygems.org`, so there's chances that your `Gemfile.lock` file is somehow corrupted and the actionable gem is not under the proper section.

This one could be similar to https://github.com/rubygems/rubygems/issues/5067.

**Comment by coldnebo** [comment]
Oh! Indeed, the lock file is structured in a weird way... all these gems aren't listed with a source.  Is that something that could happen from an old copy of bundler?  This is old code being updated.  Interesting, I'm removing the lock file and regenerating it just to see what the structure should be and then restoring the original versions.  

Sorry, not the same issue at all then! Thanks!

**Comment by deivid-rodriguez** [comment]
I'm not sure what the cause could be... In #5067 there's a similar situation created by Bundler itself, after a specific edit to the `Gemfile`, I still need to wrap up the fix for that.

**Comment by deivid-rodriguez** [comment]
#5525 should fix this!

**Comment by deivid-rodriguez** [comment]
This turned out to be a regression introduced by https://github.com/rubygems/rubygems/commit/565549260be50a84a801093ba44e399f2fafb559. Now you will get the good old message one would get before the bug was introduced, something like

```
$ bundle
Fetching gem metadata from https://rubygems.org/.........
activesupport-7.0.2.3 requires ruby version >= 2.7.0, which is incompatible with the current version, ruby 2.6.9p207
```

**Comment by ShockwaveNN** [comment]
Thanks, I'll recheck as soon as fixed version is released

**Comment by ShockwaveNN** [comment]
Can confirm fix in bundle 2.3.15

Now error is clear:
```
parallel_tests-3.8.0 requires ruby version >= 2.7.0, which is incompatible with
the current version, ruby 2.6.10p210
```

**Comment by deivid-rodriguez** [comment]
As a note, #5784 will further improve this. Now, instead of raising an error in this case, Bundler will re-resolve and update the lockfile with the newest version that supports the running Ruby. In this case, it will automatically update the lockfile to use parallel_tests 3.7.0.
